### PR TITLE
Added new sample source code. Clean up.

### LINKED ISSUE #19674: Comments say "// TODO: set large fields to null." - why?
If there's a good reason for it, please document it, otherwise please remove the comments from this document and from any auto-generated code. It's clearly confusing people as you can see if you search online for an answer to it.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 572ef6c6-5066-4a14-76a0-c3fb4b62e604
* Version Independent ID: 9c754da1-c234-fb12-580b-0fcf9ffe2c20
* Content: [Implement a Dispose method](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose)
* Content Source: [docs/standard/garbage-collection/implementing-dispose.md](https://github.com/dotnet/docs/blob/master/docs/standard/garbage-collection/implementing-dispose.md)
* Product: **dotnet**
* Technology: **dotnet-standard**
* GitHub Login: @gewarren
* Microsoft Alias: **gewarren**

### LINKED ISSUE #20354: IAsyncDispose not clear when using mixing with owned resources that implement IDisposable only
The article / examples are not clear how to handle the case where you have multiple managed resources, but some implement both `IAsyncDisposable / IDisposable` and some only implement `IDisposable`.

For example suppose I'm implementing a custom `Stream`, that needs to wrap other diposable resources.  The base class is part of .net framework, so I don't have the option of simply omitting `IAsyncDisposable`.  Further suppose that I also don't control the code of any of the resources I rely on, so modifying the interfaces is not an option.

```csharp

//I don't control this code
public class SomeDisposableResource : IDisposable
{
    ...
]

//I don't control this code
public class SomeAsyncDisposableResource : IDisposable, IAsyncDisposable
{
  ....
}

//I only control this code in the derived class.  Stream implements both IDisposable and IAsyncDisposable
public class MyCustomStream : Stream
{
     private SomeDisposableResource someDisposableResource = new SomeDisposableResource();
     private SomeAsyncDisposableResource someAsyncDisposableResource = new SomeAsyncDisposableResource();

     //What is the proper way to dispose of both someDisposableResource  and someAsyncDisposableResource 
     //for the IDisposable and IAsyncDisposable methods?
     //Note also (probably because it is abstract) that Stream does not provide DisposeAsyncCore
}

```

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 61a961dc-3ebf-2a96-a6eb-2e0b62ab909f
* Version Independent ID: d72276e1-9059-815f-dc14-a03a7544ab87
* Content: [Implement a DisposeAsync method](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync)
* Content Source: [docs/standard/garbage-collection/implementing-disposeasync.md](https://github.com/dotnet/docs/blob/master/docs/standard/garbage-collection/implementing-disposeasync.md)
* Product: **dotnet**
* Technology: **dotnet-standard**
* GitHub Login: @IEvangelist
* Microsoft Alias: **dapine**

### PR DESCRIPTION
**Pull request by IEvangelist**
## Summary

- Added project to help build validation along
- Clean up source code, adjustments per Jeremy Barton
- Add clarifying statements about assigning to `null` for large fields
- Minimize code comments to avoid non-loc
- Fixed bugs in code
- Added namespaces where appropriate
- Added new example source code for `CustomDisposable` in direct response to #20354

Fixes #19674, fixes #20354

### CODE REVIEW

File: samples/snippets/csharp/VS_Snippets_CLR/conceptual.asyncdisposable/dispose-and-disposeasync.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
+﻿using System;
```
Comment (Youssef1313) [inline samples/snippets/csharp/VS_Snippets_CLR/conceptual.asyncdisposable/dispose-and-disposeasync.cs:1]: Consider moving this file under the article directory.

Comment (IEvangelist) [reply samples/snippets/csharp/VS_Snippets_CLR/conceptual.asyncdisposable/dispose-and-disposeasync.cs:1]: I did consider it, and decided to leave it for now.

File: docs/standard/garbage-collection/implementing-dispose.md [hunk 2]
```diff
@@ -72,7 +72,7 @@ The body of the method consists of two blocks of code:
 
   - **Managed objects that implement <xref:System.IDisposable>.** The conditional block can be used to call their <xref:System.IDisposable.Dispose%2A> implementation (cascade dispose). If you have used a derived class of <xref:System.Runtime.InteropServices.SafeHandle?displayProperty=nameWithType> to wrap your unmanaged resource, you should call the <xref:System.Runtime.InteropServices.SafeHandle.Dispose?displayProperty=nameWithType> implementation here.
 
-  - **Managed objects that consume large amounts of memory or consume scarce resources.** Assign large managed object references to `null` to make them more likely to be unreachable. This releases them faster than if they were reclaimed non-deterministically.
+  - **Managed objects that consume large amounts of memory or consume scarce resources.** Assign large managed object references to `null` to make them more likely to be unreachable. This releases them faster than if they were reclaimed non-deterministically, and this is usually done outside of the conditional block.
```
Comment (tdykstra) [inline docs/standard/garbage-collection/implementing-dispose.md:75]: > the conditional block

What block is this referring to?

File: docs/standard/garbage-collection/implementing-disposeasync.md [hunk 3]
```diff
@@ -66,6 +66,18 @@ All non-sealed classes should be considered a potential base class, because they
 
 The preceding example uses the <xref:System.Text.Json.Utf8JsonWriter>. For more information about `System.Text.Json`, see [How to migrate from Newtonsoft.Json to System.Text.Json](../serialization/system-text-json-migrate-from-newtonsoft-how-to.md).
 
+## Implement both dispose and async dispose patterns
+
+You may need to implement both the <xref:System.IDisposable> and <xref:System.IAsyncDisposable> interfaces, especially when your class scope contains instances of these implementations. Doing so ensures that you can properly cascade clean up calls. Here is an example class that implements both interfaces, and demonstrates the proper guidance for cleanup.
```
Comment (tdykstra) [inline docs/standard/garbage-collection/implementing-disposeasync.md:71]:
```suggestion
You may need to implement both the <xref:System.IDisposable> and <xref:System.IAsyncDisposable> interfaces, especially when your class scope contains instances of these implementations. Doing so ensures that you can properly cascade clean up calls. Here is an example class that implements both interfaces and demonstrates the proper guidance for cleanup.
```

File: docs/standard/garbage-collection/implementing-disposeasync.md [hunk 4]
```diff
@@ -66,6 +66,18 @@ All non-sealed classes should be considered a potential base class, because they
...
+
+:::code language="csharp" source="../../../samples/snippets/csharp/VS_Snippets_CLR/conceptual.asyncdisposable/dispose-and-disposeasync.cs":::
+
+The <xref:System.IDisposable.Dispose?displayProperty=nameWithType> and <xref:System.IAsyncDisposable.DisposeAsync?displayProperty=nameWithType> implementations are both simple boilerplate code. The `Dispose(bool)` and `DisposeAsyncCore()` methods start by checking if `_disposed` is `true`, and will only run when `false`.
```
Comment (tdykstra) [inline docs/standard/garbage-collection/implementing-disposeasync.md:75]:
```suggestion
The <xref:System.IDisposable.Dispose?displayProperty=nameWithType> and <xref:System.IAsyncDisposable.DisposeAsync?displayProperty=nameWithType> implementations are both simple boilerplate code. The `Dispose(bool)` and `DisposeAsyncCore()` methods start by checking if `_disposed` is `true`, and will only run when it's `false`.
```

**Review by tdykstra — APPROVED** [review]
LGTM, one question.

File: docs/standard/garbage-collection/implementing-dispose.md (hunk 2 above)
Comment (IEvangelist) [reply docs/standard/garbage-collection/implementing-dispose.md:75]: This is a bullet for the conditional block in question, contextually it makes sense.

**Comment by imba-tjd** [comment]
How about change `if xxx is not null` to `?.`

**Pull request merged**